### PR TITLE
Adding Writeable crate and trait.

### LINKED ISSUE #181: Writeable trait for ICU4X
When implementing FixedDecimal, I [discovered](https://unicode-org.slack.com/archives/CUZ62M47Q/p1594951694008700) that, relative to std::fmt::Display, we can increase to-string performance by a considerable margin by pre-allocating string capacity.

Since ICU4X will require a lot of to-string conversion, I was wondering whether we would consider it a good idea to make a new trait similar to std::fmt::Display for use by ICU4X types.

Something like:

```rust
trait Writeable {
    fn write_to(&self, sink: &mut dyn fmt::Write) -> fmt::Result;
    fn write_len(&self) -> usize;

    fn to_string(&self) -> String {
        let mut output = String::with_capacity(self.write_len());
        self.write_to(output).expect("impl Write for String is infallible");
        Ok(output)
    }
}

// Auto-implement std::fmt::Display
impl<T: Writeable> fmt::Display for T {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.write_to(f)
    }
}
```

@Manishearth @hsivonen @zbraniecki

### PR DESCRIPTION
**Pull request by sffc**
See #181

I'm mostly happy with this, but I wasn't able to `impl<T: Writeable> ToString for T`; that gave me [E0210](https://doc.rust-lang.org/error-index.html#E0210).  ~So instead, I did `impl ToString for dyn Writeable`, which gets the job done, but it seems like you have to explicitly cast to `&dyn Writeable` whenever you want to call `to_string()`.  I'm not a stickler for the ergonomics, but I don't have a better solution.  The other potential downside of this approach is that it probably forces the `to_string()` implementation to be carried whenever a `&dyn Writeable` trait object is used, which might have a small impact on code size.~

Update 2020-11-10: I'm now using a separate method `writeable_to_string` as suggested by @Manishearth.

Benchmarks are good:

```
writeable/overview      time:   [76.000 ns 76.194 ns 76.398 ns]                               
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe

writeable/short/to_string                                                                             
                        time:   [24.418 ns 24.494 ns 24.574 ns]
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

writeable/medium/to_string                                                                             
                        time:   [24.737 ns 24.821 ns 24.943 ns]
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  7 (7.00%) high severe

writeable/long/to_string                                                                             
                        time:   [27.090 ns 27.248 ns 27.467 ns]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

display/short/to_string time:   [56.480 ns 56.767 ns 57.132 ns]                                    
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

display/medium/to_string                                                                             
                        time:   [39.795 ns 39.895 ns 40.002 ns]

display/long/to_string  time:   [41.193 ns 41.353 ns 41.539 ns]                                    
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe
```

### CODE REVIEW

File: utils/writeable/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+use std::string::ToString;
+
+/// Writeable is an alternative to std::fmt::Display with the addition of a length function.
+pub trait Writeable {
+    /// Writes bytes to the given sink. Errors from the sink are bubbled up.
+    fn write_to(&self, sink: &mut dyn fmt::Write) -> fmt::Result;
+
+    /// Returns an estimation of the number of bytes that will be written to the sink. The actual
+    /// number of bytes may be slightly different than what this function returns.
+    fn write_len(&self) -> usize;
```
Comment (zbraniecki) [inline utils/writeable/src/lib.rs:31]: Would it make sense to make it an `Option<usize>` to support use cases where the size is not known?

**Review by zbraniecki — DISMISSED** [review]
lgtm!

File: utils/writeable/src/lib.rs (hunk 1 above)
Comment (sffc) [reply utils/writeable/src/lib.rs:31]: An Option is too coarse I think.  An Enumeration would be more expressive.  I made a follow-up issue: https://github.com/unicode-org/icu4x/issues/370

File: utils/writeable/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    /// Returns an estimation of the number of bytes that will be written to the sink. The actual
+    /// number of bytes may be slightly different than what this function returns.
+    ///
+    /// This function may return an enumeration in the future. See:
+    /// https://github.com/unicode-org/icu4x/issues/370
+    fn write_len(&self) -> usize;
+}
+
+impl ToString for dyn Writeable {
```
Comment (Manishearth) [inline utils/writeable/src/lib.rs:37]: This is useful but as you note we need to explicitly cast to `dyn Writeable`, which isn't great.

What if we provide a defaulted `.writeable_to_string()` method on `Writeable` that does this, and if you want the fast ToString you call `.writeable_to_string()` instead?

**Review by Manishearth — DISMISSED** [review]
Haven't looked at the benches, but the trait itself looks okay. I wish there were better options here, I might have some ideas.

**Comment by sffc** [comment]
I created the new `writeable_to_string` method.  I get a modest performance improvement with it, down to 63.218 ns from 76.194 ns for the overview benchmark.

**Review by zbraniecki — APPROVED** [review]
lgtm!

**Review by Manishearth — APPROVED** [review]

**Pull request merged**